File: desktop-windows/src/sync/SyncClient.js
'use strict';

const fs = require('fs');
const path = require('path');

/**
 * T047 — Client de synchronisation vers le backend GUARD (mêmes 3 endpoints
 * que l'agent Android, cf. contracts/endpoint-sync-api.yaml). Offline-first
 * (Principe I) : aucune méthode ne lève d'exception fatale sur un échec
 * réseau — un échec laisse simplement la file locale intacte pour une
 * reprise ultérieure.
 *
 * Choix d'implémentation : `fetch` natif (disponible nativement depuis
 * Node 18 / Electron ≥ 28, cf. package.json `electron@^31`), pas de
 * dépendance HTTP supplémentaire.
 *
 * Persistance locale : un simple fichier JSON (pas de dépendance
 * `electron-store`, pour garder l'empreinte minimale — cohérent avec
 * SC-009). Le fichier contient uniquement des identifiants opaques
 * (terminal_id, access_token) et la file d'événements en attente ; il n'est
 * pas destiné à contenir de secret à haute valeur (pas de mot de passe), un
 * chiffrement au repos pourra être ajouté en durcissement (T053).
 */
class SyncClient {
  constructor({ baseUrl, cheminEtat }) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.cheminEtat = cheminEtat;
    this._chargerEtat();
  }

  _chargerEtat() {
    try {
      const brut = fs.readFileSync(this.cheminEtat, 'utf8');
      this.etat = JSON.parse(brut);
    } catch {
      this.etat = { terminalId: null, accessToken: null, versionSignatures: null, fileEvenements: [] };
    }
  }

  _sauvegarderEtat() {
    fs.mkdirSync(path.dirname(this.cheminEtat), { recursive: true });
    fs.writeFileSync(this.cheminEtat, JSON.stringify(this.etat, null, 2), 'utf8');
  }

  /** Appelé une fois à l'installation, avec l'organisation_token fourni à l'organisation cliente. */
  async enregistrer(deviceHash, appVersion, organisationToken) {
    try {
      const reponse = await fetch(`${this.baseUrl}/terminals/register`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          device_hash: deviceHash,
          platform: 'windows',
          app_version: appVersion,
          organisation_token: organisationToken,
        }),
      });

      if (!reponse.ok) return false;

      const corps = await reponse.json();
      this.etat.terminalId = corps.id;
      this.etat.accessToken = corps.access_token;
      this._sauvegarderEtat();

      return true;
    } catch {
      // Pas de réseau à l'installation : l'agent reste inopérant tant que
      // l'enrôlement n'a pas réussi (aucun scan_event ne peut être envoyé
      // sans access_token), mais ne plante jamais l'application.
      return false;
    }
  }

  /** Récupère et applique le delta de signatures (EF-EP-06/27) — tolère l'absence de réseau. */
  async synchroniserSignatures() {
    if (!this.etat.accessToken) return false;

    try {
      const params = this.etat.versionSignatures ? `?from_version=${encodeURIComponent(this.etat.versionSignatures)}` : '';
      const reponse = await fetch(`${this.baseUrl}/signatures/delta${params}`, {
        headers: { Authorization: `Bearer ${this.etat.accessToken}` },
      });

      if (reponse.status === 204) return true; // déjà à jour
      if (!reponse.ok) return false;

      const corps = await reponse.json();
      this.etat.versionSignatures = corps.version;
      this._sauvegarderEtat();

      return { hashesAjoutes: corps.hashes_added, reglesYaraAjoutees: corps.yara_rules_added };
    } catch {
      return false;
    }
  }

  /** Ajoute un événement à la file locale (appelé par le pipeline de scan, jamais bloquant sur le réseau). */
  enfilerEvenement(evenement) {
    this.etat.fileEvenements.push(evenement);
    this._sauvegarderEtat();
  }

  /** Envoie la file locale de scan-events non encore synchronisés, en un seul lot. */
  async synchroniserEvenements() {
    if (!this.etat.accessToken || !this.etat.terminalId) return false;
    if (this.etat.fileEvenements.length === 0) return true;

    try {
      const reponse = await fetch(`${this.baseUrl}/scan-events`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.etat.accessToken}`,
        },
        body: JSON.stringify({
          terminal_id: this.etat.terminalId,
          events: this.etat.fileEvenements,
        }),
      });

      // 202 (accepté) ou 409 (déjà connu du backend, idempotence par
      // client_event_id) marquent la file comme traitée : dans les deux cas,
      // retenter n'apporterait rien de plus.
      if (reponse.ok || reponse.status === 409) {
        this.etat.fileEvenements = [];
        this._sauvegarderEtat();
        return true;
      }

      return false;
    } catch {
      return false;
    }
  }
}

module.exports = { SyncClient };

File: mobile-android/app/src/main/kotlin/africa/trugroup/guard/sync/SyncClient.kt
package africa.trugroup.guard.sync

import africa.trugroup.guard.data.GuardDatabase
import africa.trugroup.guard.data.HashConnuEntity
import africa.trugroup.guard.data.RegleYaraEntity
import africa.trugroup.guard.data.ScanEventEntity
import africa.trugroup.guard.data.VersionSignaturesLocaleEntity
import android.content.Context
import android.content.SharedPreferences

/**
 * T042 — Orchestration de la synchronisation offline-first (Principe I) :
 * ne lève jamais d'exception fatale sur un échec réseau, se contente de
 * laisser la file locale intacte pour une reprise ultérieure.
 */
class SyncClient(
    private val context: Context,
    private val api: GuardApi,
    private val prefs: SharedPreferences,
) {
    private val db get() = GuardDatabase.obtenir(context)

    /** Appelé une fois à l'installation, avec le organisation_token fourni à l'organisation cliente. */
    suspend fun enregistrer(deviceHash: String, plateforme: String, versionApp: String, organisationToken: String): Boolean {
        val reponse = runCatching {
            api.enregistrerTerminal(
                RegisterTerminalRequest(deviceHash, plateforme, versionApp, organisationToken),
            )
        }.getOrNull() ?: return false

        val corps = reponse.body() ?: return false
        if (! reponse.isSuccessful) return false

        prefs.edit()
            .putString(CLE_TERMINAL_ID, corps.id)
            .putString(CLE_ACCESS_TOKEN, corps.accessToken)
            .apply()

        return true
    }

    /** Récupère et applique le delta de signatures (EF-EP-06/27) — tolère l'absence de réseau. */
    suspend fun synchroniserSignatures(): Boolean {
        val token = prefs.getString(CLE_ACCESS_TOKEN, null) ?: return false
        val versionLocale = db.signatureDao().versionActuelle()

        val reponse = runCatching {
            api.deltaSignatures("Bearer $token", versionLocale)
        }.getOrNull() ?: return false

        if (reponse.code() == 204) return true // déjà à jour
        val corps = reponse.body() ?: return false
        if (! reponse.isSuccessful) return false

        db.signatureDao().ajouterHashes(corps.hashesAdded.map { HashConnuEntity(it) })
        db.signatureDao().ajouterRegles(corps.yaraRulesAdded.map { RegleYaraEntity(texteYar = it) })
        db.signatureDao().definirVersion(VersionSignaturesLocaleEntity(numeroVersion = corps.version))

        return true
    }

    /** Envoie la file locale de scan-events non encore synchronisés (T041), en un seul lot. */
    suspend fun synchroniserEvenements(): Boolean {
        val token = prefs.getString(CLE_ACCESS_TOKEN, null) ?: return false
        val terminalId = prefs.getString(CLE_TERMINAL_ID, null) ?: return false
        val enAttente = db.scanEventDao().enAttenteDeSynchro()
        if (enAttente.isEmpty()) return true

        val requete = ScanEventsRequest(
            terminalId = terminalId,
            events = enAttente.map {
                ScanEventPayload(
                    clientEventId = it.clientEventId,
                    trigger = it.trigger,
                    sha256 = it.sha256,
                    yaraRuleMatched = it.yaraRuleMatched,
                    levelReached = it.levelReached,
                    classification = it.classification,
                    actionTaken = it.actionTaken,
                    occurredAt = it.occurredAtIso8601,
                )
            },
        )

        val reponse = runCatching { api.envoyerEvenements("Bearer $token", requete) }.getOrNull() ?: return false

        // 202 (accepté) ou 409 (déjà connu du backend) marquent la file comme traitée :
        // dans les deux cas, retenter n'apporterait rien de plus.
        if (reponse.isSuccessful || reponse.code() == 409) {
            db.scanEventDao().marquerSynchronise(enAttente.map { it.clientEventId })

            return true
        }

        return false
    }

    /** Ajoute un événement à la file locale (appelé par le pipeline de scan, jamais bloquant sur le réseau). */
    suspend fun enfilerEvenement(evenement: ScanEventEntity) {
        db.scanEventDao().enfiler(evenement)
    }

    companion object {
        private const val CLE_TERMINAL_ID = "terminal_id"
        private const val CLE_ACCESS_TOKEN = "access_token"
    }
}
